prow/plugins/configtree.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugins
+
+type ProwConfig interface {
+	Apply(config ProwConfig) ProwConfig
+}
+
+// ConfigTree specifies the global generic config for a plugin.
+type ConfigTree[T ProwConfig] struct {
+	Config T
```

Comment:
one issue with generics is that we cannot embed T:
`embedded field type cannot be a (pointer to a) type parameter`